src/parser.js

```diff
@@ -1892,6 +1892,9 @@ export class GenericParser extends Tokenizer {
             rest = expr;
             break;
           }
+          if (expr.type !== 'ArrayExpression' && expr.type !== 'ObjectExpression' && !isValidSimpleAssignmentTarget(expr)) {
```

Comment:
The only thing I can think of such that this criteria is satisfied but we would not already have recognized that we can't be parsing an assignment target is `AssignmentExpression`, so it's possible we could replace this condition with `if (expr.type === 'AssignmentExpression')`.

On the other hand, this is closer to [what the grammar says](https://tc39.github.io/ecma262/#sec-destructuring-assignment-static-semantics-early-errors).